### PR TITLE
Clean up conversation view.

### PR DESCRIPTION
**Pull request by charlesmchen**
This PR finalizes a bunch of changes from the JSQ rewrite.

PTAL @michaelkirk

### CODE REVIEW

File: Signal/src/ViewControllers/ConversationView/ConversationInputTextView.h [hunk 1]
```diff
@@ -10,8 +10,6 @@ NS_ASSUME_NONNULL_BEGIN
 
 - (void)didPasteAttachment:(SignalAttachment *_Nullable)attachment;
 
-- (void)textViewDidChangeLayout;
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/ConversationInputTextView.h]: We don't need to observe the layout of the input toolbar or text view anymore; it's redundant with and less useful than our new approach of observing layout changes to the collection view.

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 2]
```diff
@@ -606,8 +609,6 @@ - (void)scrollToDefaultPosition
     } else {
         [self scrollToBottomAnimated:NO];
     }
-
-    [self updateLastVisibleTimestamp];
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: We need a thorough test plan around this, but think we were updating the "last visible" state way too often.  It should be sufficient to update it:

* On view will appear.
* When collection view contents are changed with `performBatchUpdates`.
* When collection view scroll state changes.

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 3]
```diff
@@ -2239,7 +2226,6 @@ - (void)ensureScrollDownButton
     OWSAssert([NSThread isMainThread]);
 
     BOOL shouldShowScrollDownButton = NO;
-    NSUInteger numberOfMessages = [self.messageMappings numberOfItemsInSection:0];
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: We want to treat the "view items" as the single source of truth around the collection view contents and consult the `messageMappings` only when updating the "view items".

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 4]
```diff
@@ -2248,12 +2234,11 @@ - (void)ensureScrollDownButton
     // one page.
     BOOL isScrolledUp = scrollSpaceToBottom > pageHeight * 1.f;
 
-    if (numberOfMessages > 0) {
-        TSInteraction *lastInteraction =
-            [self interactionAtIndexPath:[NSIndexPath indexPathForRow:(NSInteger)numberOfMessages - 1 inSection:0]];
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: Similarly, we want to stop using `interactionAtIndexPath` in favor of the view items.

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 5]
```diff
@@ -3184,42 +3163,11 @@ - (void)attachmentButtonPressed
...
-    ConversationViewItem *_Nullable viewItem = [self viewItemForIndex:(NSUInteger)indexPath.row];
-    return viewItem.interaction;
-}
-
 - (void)updateLastVisibleTimestamp
 {
-    TSInteraction *lastVisibleInteraction = [self lastVisibleInteraction];
-    if (lastVisibleInteraction) {
-        uint64_t lastVisibleTimestamp = lastVisibleInteraction.timestampForSorting;
+    ConversationViewItem *_Nullable lastViewItem = [self.viewItems lastObject];
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: How do we know that the `self.viewItems.lastObject` is visible?

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 6]
```diff
@@ -3994,13 +3922,15 @@ - (void)reloadInteractionForViewItem:(ConversationViewItem *)viewItem
...
-- (nullable ConversationViewItem *)viewItemForIndex:(NSUInteger)index
+- (nullable ConversationViewItem *)viewItemForIndex:(NSInteger)index
 {
-    if (index >= self.viewItems.count) {
+    if (index < 0 || index >= (NSInteger)self.viewItems.count) {
         OWSFail(@"%@ Invalid view item index: %zd", self.tag, index);
         return nil;
     }
-    return self.viewItems[index];
+    ConversationViewItem *_Nullable viewItem = self.viewItems[(NSUInteger)index];
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: When would this be nil?

Comment (charlesmchen) [reply Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: You're right; this is redundant with the `OWSFail()` above. Removed.

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m (hunk 5 above)
Comment (charlesmchen) [reply Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: Good catch, this is wrong. Opening a new PR.

Sorry, I missed your comments last night.

**Pull request merged**